scripts/helper/find-idents.c

```diff
@@ -0,0 +1,497 @@
+ */
+
+#define _POSIX_C_SOURCE 200809L
+#include <assert.h>
+#include <ctype.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+struct buf {
+	char *buf;
+	int alloc;
+	int len;
+};
+
+enum token_kind {
+	TOKEN_NONE = 0,
+	TOKEN_IDENTIFIER, /* For example: static extern if while */
+	TOKEN_LITERAL, /* For example: 0xff 42 "foo" */
+	TOKEN_SPECIAL, /* For example: ++ -= ! ... */
+};
+
+struct token {
+	int line;
+	enum token_kind kind;
+	struct buf name;
+	unsigned int special;
+};
+
+enum {
+	SPECIAL_ELLIPSIS = 256,
+	SPECIAL_ASSIGN,
+	SPECIAL_BIT_OP,
+	SPECIAL_INC_OP,
+	SPECIAL_DEC_OP,
+	SPECIAL_PTR_OP,
+	SPECIAL_AND_OP,
+	SPECIAL_OR_OP,
+	SPECIAL_COMPARISON_OP,
+	SPECIAL_COMMENT_BEGIN,
+	SPECIAL_COMMENT_END,
+	SPECIAL_COMMENT_LINE_BEGIN,
+};
+
+static char *current_buffer_position;
+static struct token *tokens;
+static int nr_tokens, alloc_tokens;
+static int current_line;
+static char **argv_tokens;
+static int found_token;
+
+static const char find_banned_usage[] =
+"Usage: find-banned [OPTIONS...] FILE\n"
+"When FILE is -, read stdin\n"
+"OPTIONS:\n"
+"  --tokens=<tokens>     Comma-separated string of idents to grep for\n";
+
+static void
+usage(void)
+{
+	printf("%s", find_banned_usage);
+	exit(0);
+}
+
+char **
+split(char *str, char delim)
+{
+	if (!str) {
+		return NULL;
+	}
+	int argc = 1;
+	char *p = str;
+	while (*p) {
+		if (*p == delim) {
+			argc++;
+		}
+		++p;
+	}
+
+	char **argv = calloc(argc + 1, sizeof(*argv));
+	char **argvp = argv;
+	p = str;
+	while (*str) {
+		if (*str == delim) {
+			*argvp++ = strndup(p, str-p);
+			p = str + 1;
+		}
+		++str;
+	}
+	*argvp++ = strndup(p, str-p);
+	*argvp = NULL;
+	return argv;
+}
+
+void
+buf_init(struct buf *s)
+{
+	s->alloc = 256;
+	s->buf = malloc(s->alloc);
+	s->buf[0] = '\0';
+	s->len = 0;
+}
+
+void
+buf_add(struct buf *s, const char *data, size_t len)
+{
+	if (!data || data[0] == '\0') {
+		return;
+	}
+	if (s->alloc <= s->len + len + 1) {
+		s->alloc = s->alloc + len;
+		s->buf = realloc(s->buf, s->alloc);
+	}
+	memcpy(s->buf + s->len, data, len);
+	s->len += len;
+	s->buf[s->len] = 0;
+}
+
+void
+buf_add_char(struct buf *s, char ch)
+{
+	if (s->alloc <= s->len + 1) {
+		s->alloc = s->alloc * 2 + 16;
+		s->buf = realloc(s->buf, s->alloc);
+	}
+	s->buf[s->len++] = ch;
+	s->buf[s->len] = 0;
+}
+
+static struct token *
+add_token(void)
+{
+	if (nr_tokens == alloc_tokens) {
+		alloc_tokens = (alloc_tokens + 16) * 2;
+		tokens = realloc(tokens, alloc_tokens * sizeof(struct token));
+	}
+	struct token *token = tokens + nr_tokens;
+	memset(token, 0, sizeof(*token));
+	nr_tokens++;
+	buf_init(&token->name);
+	token->line = current_line;
+	return token;
+}
+
+static void
+handle_whitespace(struct token *token)
+{
+	return;
+	if (current_buffer_position[0] == '\n') {
+		++current_line;
+	}
+	current_buffer_position++;
+	if (isspace(current_buffer_position[0])) {
+		handle_whitespace(token);
+	}
+}
```

Comment:
This function is stubbed out. I think this came from my patch. We should likely either remove the function completely or fix it some other way. See top message of https://github.com/johanmalm/labwc/pull/8#issue-1577310757